Polymorphic Scale Client

DirectXMan12 opened the pull request:
This PR introduces a polymorphic scale client based on discovery information that's able to scale scalable resources in arbitrary group-versions, as long as they present the scale subresource in their discovery information.

Currently, it supports `extensions/v1beta1.Scale` and `autoscaling/v1.Scale`, but supporting other versions of scale if/when we produce them should be fairly trivial.

It also updates the HPA to use this client, meaning the HPA will now work on any scalable resource, not just things in the `extensions/v1beta1` API group.

**Release note**:
```release-note
Introduces a polymorphic scale client, allowing HorizontalPodAutoscalers to properly function on scalable resources in any API group.
```

Unblocks #29698
Unblocks #38756
Unblocks #49504 
Fixes #38810

DirectXMan12 committed:
[client-go] avoid Registry in fake REST client

Previously, the fake RESTClient in client-go required a Registry.  It
used the Registry to fetch the GroupVersion for the fake client.
However, the way it did so was dubious in some cases (it hard-coded the
default API group in places), and not strictly necssary.

This updates the fake client to just recieve the GroupVersion and
internal group name directly, instead of requiring a Registry, so that
it can be consumed in unit tests where a Registry isn't necessarily
readily available (e.g. elsewhere in client-go).

DirectXMan12 committed:
[client-go] fake discovery returns server groups

The fake discovery client currently returns `nil, nil` for several
methods.  Among them is the `ServerGroups` method, which is used by the
discovery REST mapper implementations.  This updates the fake discovery
client to actually return server groups so that the discovery REST
mapper can be used in tests.

DirectXMan12 committed:
[client-go] Polymorphic Scale Client

This introduces a polymorphic scale client capable of operating against
scale subresources which return different group-versions of Scale.  The
scale subresources may be in group-versions different than the scale
itself, so that we no longer need a copy of every scalable resource in
the extensions API group.

To discovery which Scale group-versions go to which subresources,
discovery is used.

The scale client maintains its own internal versions and conversions to
several external versions, with a "hub" version that's a copy of the
autoscaling internal version.

It currently supports the following group-versions for Scale subresources:

- extensions/v1beta1.Scale
- autoscaling/v1.Scale

DirectXMan12 committed:
Make HPA controller use polymorphic scale client

This updates the HPA controller to use the polymorphic scale client from
client-go.  This should enable HPAs to work with arbitrary scalable
resources, instead of just those in the extensions API group (meaning we
can deprecate the copy of ReplicationController in extensions/v1beta1).
It also means that the HPA controller now pays attention to the
APIVersion field in `scaleTargetRef` (more specifically, the group part
of it).

Note that currently, discovery information on which resources are
available where is only fetched once (the first time that it's
requested).  In the future, we may want a refreshing discovery REST
mapper.

DirectXMan12 commented:
cc @deads2k @kargakis @soltysh at long last ;-)

DirectXMan12 commented:
P.S. I think I got most of the open issues around this, but I'm sure there are more.  Let me know if I missed one.

deads2k commented:
@sttts

deads2k commented on cmd/kube-controller-manager/app/autoscaling.go:
```diff
@@ -21,7 +21,12 @@ limitations under the License.
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	discocache "k8s.io/client-go/discovery/cached" // Saturday Night Fever
```
if the rest of the this pull earns it, I'll let you have it :)

DirectXMan12 replied:
whoops, didn't mean to leave that in :-P

deads2k commented on staging/src/k8s.io/client-go/scale/util.go:
```diff
@@ -0,0 +1,156 @@
+	r.mu.RLock()
+	gv, exists := r.subresMap[resource]
+	if !exists {
+		// retry on misses in case we have an update
```
Can you build a non-caching version, then build a caching version on top of it?  This is a little hard for me to read and the complexity can be built up in another pull.

DirectXMan12 replied:
sure, can-do

DirectXMan12 replied:
hmm... this is a little weirder than I initially thought, since a non-cached version would probably use different code paths entirely (i.e. it wouldn't use the "get-all-at-once" API).

deads2k replied:
> hmm... this is a little weirder than I initially thought, since a non-cached version would probably use different code paths entirely (i.e. it wouldn't use the "get-all-at-once" API).

Yeah, that's the bit I don't like.

deads2k commented on staging/src/k8s.io/client-go/scale/util.go:
```diff
@@ -0,0 +1,156 @@
+	if !exists {
+		// retry on misses in case we have an update
+		r.mu.RUnlock()
+		r.generateKindMap()
```
ignoring an error doesn't look right.

deads2k commented on staging/src/k8s.io/client-go/scale/util.go:
```diff
@@ -0,0 +1,156 @@
+}
+
+func NewScaleConverter() *ScaleConverter {
+	scheme := runtime.NewScheme()
```
This makes me very happy.

deads2k commented on staging/src/k8s.io/client-go/scale/util.go:
```diff
@@ -0,0 +1,156 @@
+		scheme: scheme,
+		internalVersioner: runtime.NewMultiGroupVersioner(
+			scalescheme.SchemeGroupVersion,
+			schema.GroupKind{Group: scaleext.GroupName, Kind: "Scale"},
```
This ordering controls default preference, right?  Make autoscaling come first.

DirectXMan12 replied:
nope, the order shouldn't matter -- this says "if any of the objects passed in match autoscaling/*.Scale or extensions/*.Scale, return autoscaling/__scale_internal.Scale as the target GVK"

deads2k commented on staging/src/k8s.io/client-go/scale/scheme/types.go:
```diff
@@ -0,0 +1,60 @@
+)
+
+// This file contains our own "internal" version of scale that we use for conversions,
+// since we can't use the main Kubernetes internal versions.
```
This makes me very sad, but so be it.

caesarxuchao replied:
IIUC with the "alternative API representations", client can request the server to do the conversion. So the internal types and the entire conversion machinery can be removed in the future. Am I too optimistic?

deads2k replied:
> IIUC with the "alternative API representations", client can request the server to do the conversion. So the internal types and the entire conversion machinery can be removed in the future. Am I too optimistic?

Too optimistic.  This code will still be correct even once we get there.  I would guess we're multiple releases away from re-loosening the server to support this.

deads2k commented on staging/src/k8s.io/client-go/scale/scheme/register.go:
```diff
@@ -0,0 +1,53 @@
+limitations under the License.
+*/
+
+package scheme
```
@caesarxuchao have a look at the package location.  It looks ok to me.

caesarxuchao replied:
The location lgtm, too.

sttts commented:
/cc @nikhita

deads2k commented on staging/src/k8s.io/client-go/scale/scheme/extensionsv1beta1/register.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+// GroupName is the group name use in this package
+const GroupName = "extensions"
```
these have to match exactly and you already have a package dependency.  Assign them to the `extensionsapiv1beta1` values I think.

DirectXMan12 replied:
ah, good point, will do.

deads2k commented on staging/src/k8s.io/client-go/scale/scheme/types.go:
```diff
@@ -0,0 +1,60 @@
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Scale represents a scaling request for a resource.
+type Scale struct {
```
need fuzzing tests on this.

deads2k commented on staging/src/k8s.io/client-go/scale/interfaces.go:
```diff
@@ -0,0 +1,38 @@
+	Get(kind schema.GroupKind, name string) (*autoscalingapi.Scale, error)
+
+	// Update updates the scale of the the given scalable resource.
+	Update(ind schema.GroupKind, scale *autoscalingapi.Scale) (*autoscalingapi.Scale, error)
```
Why aren't you using the hub you defined?  If you want to use `autoscaling/v1` as your hub, why define a new internal to convert to?

DirectXMan12 replied:
I considered it, but I figured we didn't want to ever expose internal versions in clients.  It does make some sense to me to use it, though (although it also make some sense to use internal versions elsewhere, so /me shrugs).

On the flip side, even though we don't expose it, having an internal version makes it a bit easier to reason about the conversions (IMO), especially since it's following the normal Kubernetes hub-and-spoke pattern.  AFAIK, there's not a technical reason why we couldn't use external as the hub, though.

deads2k replied:
> On the flip side, even though we don't expose it, having an internal version makes it a bit easier to reason about the conversions (IMO), especially since it's following the normal Kubernetes hub-and-spoke pattern. AFAIK, there's not a technical reason why we couldn't use external as the hub, though.

I think this demonstrates the inherent problem of not exposing internal types.  If you want to keep your own hub, I can live with it.  You need fuzzers though.

deads2k commented on staging/src/k8s.io/client-go/scale/client.go:
```diff
@@ -0,0 +1,205 @@
+		Resource(mapping.Resource).
+		Name(name).
+		SubResource("scale").
+		Do().
```
If you switched to use a `DoRaw`, then used your custom scheme for decoding, wouldn't you be able to use any given `rest.Interface` without constructing a new client from config?  I like supporting `NewForConfig`, I'm just not seeing why I need to have multiple clients and not supporting `New`.

DirectXMan12 replied:
We still need a `rest.Interface` for each group-version, no?

caesarxuchao replied:
> We still need a rest.Interface for each group-version, no?

Is it because you need to point to different endpoints? You can construct a path and call `AbsPath()`. We can expose the function the restclient uses to build the path so that we don't have to hard code here.

DirectXMan12 replied:
Actually making `defaultServerURLFor` public seems reasonable, but just circumventing the request with `AbsPath` seems hacky.  I think if we're going to change this, we should have a type that provides clients for a given GroupVersion only, and use it in both the scale client and dynamic client (since the dynamic client currently uses the same approach that I have here).

For example:

```go
type Provider interface {
    GroupVersion(gv schema.GroupVersion) rest.Interface
}

provider.GroupVersion(schema.GroupVersion{"apps", "v1beta2"}).Get()
  .Resource("deployments")
  .Name("foo")
  .SubResource("scale")
  .Do()
```

That looks a bit clunky with the `GroupVersion` before the Get, but provides a rest.Interface out of it, so it fits with more existing code.  An alternative might be:

```go
type MultiVersionInterface interface {
    // the same methods as rest.Interface, but they return a GroupVersionProvider below
}

type GroupVersionProvider interface {
    GroupVersion(schema.GroupVersion) rest.Request
}

type client MultiVersionInterface = ...
client.Get().GroupVersion(schema.GroupVersion{"apps", "v1beta2"})
   .Resouce("deployments")
   .Name("foo")
   .Do()
```

But since that doesn't return a `rest.Interface` anywhere, it's mostly useful only here and in the dynamic client.

WDYT?

deads2k replied:
> But since that doesn't return a rest.Interface anywhere, it's mostly useful only here and in the dynamic client.
> 
> WDYT?

I think the dynamic client interface needs some work, but I would definitely accept an `.AbsPath` here to make the externally facing bits of this API easier to work with.

caesarxuchao replied:
Let's use `.AbsPath` here and refactor it with the dynamic client in another PR.

Also, with the first approach, it looks like the provider still returns different restclient for different group versions. That sounds unnecessary.

Before taking the second approach, could we try to add a method to the rest.Request to allow us change group/version after its construction?

deads2k commented on staging/src/k8s.io/client-go/scale/interfaces.go:
```diff
@@ -0,0 +1,38 @@
+// the scale subresource.
+type ScaleInterface interface {
+	// Get fetches the scale of the given scalable resource.
+	Get(kind schema.GroupKind, name string) (*autoscalingapi.Scale, error)
```
This should take a `GroupResource`.  `GroupKind` was a mistake.  The responsibility of determining which resource to use should lie with the caller.

deads2k commented on pkg/kubectl/cmd/annotate_test.go:
```diff
@@ -451,7 +451,7 @@ func TestAnnotateObject(t *testing.T) {
 	tf.Printer = &testPrinter{}
 	tf.UnstructuredClient = &fake.RESTClient{
-		APIRegistry:          api.Registry,
+		GroupVersion:         api.Registry.GroupOrDie(api.GroupName).GroupVersion,
```
Is this fix in a separate commit. You've got lots of these lines and this update looks good.  We could split it out and merge it.

DirectXMan12 replied:
yeah, it's in its own commit (https://github.com/kubernetes/kubernetes/pull/53743/commits/8f9abfd3b08c14db48c251c65edc3973658446b6)

deads2k commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -370,7 +362,15 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
+		a.updateStatusIfNeeded(hpaStatusOriginal, hpa)
+		return fmt.Errorf("invalid API version in scale target reference: %v", err)
+	}
+	targetGK := targetGV.WithKind(hpa.Spec.ScaleTargetRef.Kind).GroupKind()
```
You need to resolve a `GroupResource` here.  You client should deal in the correct types.  It helps make it clear that the HPA Kind make a mistake in their API.

DirectXMan12 replied:
ack, makes sense to me

caesarxuchao commented:
Will kubectl use this client? cc @foxish

foxish commented:
@caesarxuchao Yeah, `kubectl scale` should then be able to use this polymorphic client instead.

caesarxuchao commented:
@DirectXMan12 could you describe what an ideal world will be, supposing we have the alternative API representation, have deprecated extensions/v1beta1.Autoscale?
I think in that ideal world, the scale client will have the same interface, but it won't need to do any conversion or discovery.

deads2k commented:
> @DirectXMan12 could you describe what an ideal world will be, supposing we have the alternative API representation, have deprecated extensions/v1beta1.Autoscale?
> I think in that ideal world, the scale client will have the same interface, but it won't need to do any conversion or discovery.

I think you're correct, but I think that is a long way off.

sttts commented on federation/pkg/kubefed/unjoin_test.go:
```diff
@@ -184,9 +184,9 @@ func testUnjoinFederationFactory(name, server, secret string) cmdutil.Factory {
 	tf.Client = &fake.RESTClient{
-		APIRegistry:          api.Registry,
+		GroupVersion:         api.Registry.GroupOrDie("federation").GroupVersion,
+		InternalGroupName:    api.Registry.GroupOrDie("federation").GroupVersion.Group,
```
do we have cases with distinct internal and external group names? I think we always register the internal types into all relevant internal group versions.

DirectXMan12 replied:
I don't think so, but figured I'd let them be explicitly separate here since they were previously controlled separately.  I can change that if you want.

sttts replied:
It simplifies the interface.  So why not.

DirectXMan12 committed:
[make-rules] test grep should treat binary as text

The output of `go test` is passed throug a grep filter that's used when
producing junit output.  Unfortunately, when testing round-trip
conversions with random strings, grep can become convinced that the test
output is binary, and will truncate a failed test before any output is
displayed, showing "binary file (standard input) matches".

This forces grep to consider test output to be text, so that we always
see test results.

DirectXMan12 committed:
Fuzz label selectors

Previously, we did not have custom code for fuzzing label selectors.
Anything that used a label selector (like Scale) had to manually bypass
fuzzing the selector, or write its own fuzzer.  This introduces a fuzzer
for label selectors which generates random correct selectors with random
keys and values.

DirectXMan12 committed:
Update extensions fuzzer to use selector fuzzer

This removes the override on the extensions fuzzer so that
extensions.Scale will use the full label selector fuzzer.

DirectXMan12 committed:
[client-go] Add fake scale client

This adds a new fake scale client (for use in testing) to match the
new polymorphic scale client.

DirectXMan12 commented:
This should cover most comments and concerns.  Of note:

- We just expect client passed to `New` to have an appropriate serializer for deserializing the objects we care about.  In the case of `NewForConfig`, we set this up for you, but you could totally set it up yourself using similar methods.

- I still have to update the staging godeps, will do that shortly.

deads2k commented on hack/make-rules/test.sh:
```diff
@@ -289,7 +289,7 @@ runTests() {
      "${testargs[@]:+${testargs[@]}}" \
      | tee ${junit_filename_prefix:+"${junit_filename_prefix}.stdout"} \
-     | grep "${go_test_grep_pattern}" && rc=$? || rc=$?
+     | grep --binary-files=text "${go_test_grep_pattern}" && rc=$? || rc=$?
```
@sttts what does this do in context?

DirectXMan12 replied:
it forces grep to consider output it might otherwise think is binary as text, so that it outputs the actual match instead of just saying "binary file matched".  Since it's just test output, we should never have *actual* binary junk, just stuff that grep incorrectly assumes is binary.

DirectXMan12 replied:
I hit this when debugging a fuzzer failure -- it thought the fuzzer output was binary, and so it just truncated the results of the test output

deads2k commented on cmd/kube-controller-manager/app/autoscaling.go:
```diff
@@ -63,12 +68,29 @@ func startHPAControllerWithLegacyClient(ctx ControllerContext) (bool, error) {
+	// on cache misses.
+	cachedDiscovery := discocache.NewMemCacheClient(hpaClientGoClient.Discovery())
+	restMapper := discovery.NewDeferredDiscoveryRESTMapper(cachedDiscovery, apimeta.InterfacesForUnstructured)
+	restMapper.Reset()
```
please open an issue to put this on the controller context and share it amongst all controllers.  namespace cleanup, GC, and this come to mind.

DirectXMan12 replied:
https://github.com/kubernetes/kubernetes/issues/54084

deads2k commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -353,7 +349,28 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
+		Group: targetGV.Group,
+		Kind:  hpa.Spec.ScaleTargetRef.Kind,
+	}
+	mapping, err := a.mapper.RESTMapping(targetGK)
```
`RESTMappings` and iterate over the results until one works.  If none work, then fail with the first error.

deads2k commented on pkg/controller/podautoscaler/legacy_horizontal_test.go:
```diff
@@ -131,13 +134,11 @@ func (tc *legacyTestCase) prepareTestClient(t *testing.T) *fake.Clientset {
 			name:       "test-rc",
-			apiVersion: "extensions/v1beta1",
-			kind:       "replicationcontrollers",
+			apiVersion: "v1",
```
yay!

deads2k commented on staging/src/k8s.io/client-go/scale/client.go:
```diff
@@ -0,0 +1,194 @@
+	}
+
+	// convert this to whatever this endpoint wants
+	scaleUpdate, err := scaleConverter.ConvertToVersion(scale, desiredGVK.GroupVersion())
```
do the conversion here to set the right bytes rather than making the caller sort out a codec on the client they pass.

DirectXMan12 replied:
ack

deads2k commented on staging/src/k8s.io/client-go/scale/util.go:
```diff
@@ -0,0 +1,163 @@
+
+// ScaleKindResolver knows about the relationship between
+// resources and the GroupVersionKind of their scale subresources.
+type ScaleKindResolver interface {
```
Why is this exported?  `scaleKindResolver`?

DirectXMan12 replied:
it's passed as an argument to construct the scale client, so it needs to be public.  The default implementation is via discovery, but for testing purposes it's useful to be able to use a fake implementation.

deads2k commented on staging/src/k8s.io/client-go/scale/util.go:
```diff
@@ -0,0 +1,163 @@
+	return nil
+}
+
+func (r *discoveryScaleResolver) ScaleForResource(resource schema.GroupVersionResource) (schema.GroupVersionKind, error) {
```
could this `scaleForResource`?

DirectXMan12 commented:
For posterity: changing the code so that callers of `New` don't have to pass the correct codecs in is proving difficult.  Using `DoRaw` and encoding/decoding ourselves means duplicating a lot of logic that's in request to handle content-type negotiation, status messages, etc.  Attempting to rework `rest.Client` and/or `rest.Request` to have an escape hatch appears to be difficult, since the assumption of one set of serializers and a single group-version appear to be fairly deeply ingrained.

I'll look at fixing that in a follow-up PR.  So far, the follow-up list is

- Not requiring passing in the correct codecs
- Passing in a custom group-version to request to not require AbsPath (related to the above)

deads2k commented:
Needs to be green, but

/lgtm

deads2k commented:
@smarterclayton needs approval since we're adding a client-go dir (chao agreed on location)

smarterclayton commented:
/approve

sttts commented on staging/src/k8s.io/client-go/scale/client.go:
```diff
@@ -0,0 +1,197 @@
+import (
+	"fmt"
+
+	autoscaling "k8s.io/api/autoscaling/v1"
```
this design puts us into a strange situation when we want to add /scale support to CRDs:

1. we don't have Scale in meta/v1 because it is not general enough. Instead we put it into k8s.io/api.
2. apiextensions-apiserver is not kube-specific enough to move its types into k8s.io/api, instead we keep it logically separate.
3. client-go cannot import k8s.io/apiextensions-apiserver due to an import cycle.

Following 2, we would like to have meta/v1.Scale or apiextensions/v1beta1.Scale. But 1 and 3 stops us from that. Technically, we can import autoscaling/v1 in k8s.io/apiextensions-apiserver and use that Scale type for CRDs. But this conflicts with 2.

deads2k commented:
> this design puts us into a strange situation when we want to add /scale support to CRDs:

I'm not sure why that applies.  The apiextensions-server relies on client-go and hosts the endpoint as `Scale.v1.autoscaling`.  `client-go` can then scale custom resources without modification.

sttts commented:
> I'm not sure why that applies. The apiextensions-server relies on client-go and hosts the endpoint as Scale.v1.autoscaling. client-go can then scale custom resources without modification.

That's what I wrote: technically it's not an issue. But with the argument that apiextensions-apiserver should not move to k8s.io/api as it is not kube-related enough, it's strange to use a kube-type for scaling.

deads2k commented:
> That's what I wrote: technically it's not an issue. But with the argument that apiextensions-apiserver should not move to k8s.io/api as it is not kube-related enough, it's strange to use a kube-type for scaling.

I see that as the same argument that says, "openshift isn't kube, but we can use kube types like `Scale` for our resources".  Not everything that is scalable ends up in `k8s.io/api`.

sttts commented:
> "openshift isn't kube, but we can use kube types like Scale for our resources"

But kube wouldn't use OpenShift types. It's not a symmetric relation.

Also note that we don't have the internal type. So we have to build our own. So in the kube-apiserver binary we will have two autoscaling groups, depending on which scheme you look.

deads2k commented:
@DirectXMan12 failures look real

DirectXMan12 commented:
yep, I think something broke when I had to rebase.

DirectXMan12 commented:
ok, rebase issues should be solved

deads2k commented:
/lgtm
